sigs/jvm/rfcs/20191023-tensorflow-keras.md

```diff
@@ -0,0 +1,342 @@
+These are: 
+
+1. `Layer.build(Ops tf, Shape inputShape)`
+
+    The `build` method is intended for creating tensorflow variables which
+    will be required to compute the output of a layer. For example, for a
+    Dense layer implementation, `build` is used to create the kernel (*w_k*) and
+    bias (*w_b*) weight matrices.
+
+    To add weight variables to a layer, subclasses overriding the `build` method must call
+    the protected method `Layer.addWeight`; an internal wrapper for `tf.variable`.
+    See the Dense layer implementation for an example of this usage.
+
+    There are two build methods; one that should be overridden in any `Layer` subclass,
+    and a public method which wraps the overridden method, to be called by external
+    users of any Layer class. 
+    ```java
+    // Override this in Layer subclasses. To access dtype
+    // within this method, call this.getDtype()
+    protected abstract void build(Ops tf, Shape inputShape)
+
+    // Call this method when using a created Layer object externally.
+    // Requires the user to provide a dtype, as some tensorflow 
+    // operations require the explicit specification of a dtype
+    // (e.g. tf.variable(Shape shape, Class<T> dtype) and tf.placeholder(Class<T> dtype))
+    public final void build(Ops tf, Shape inputShape, Class<T> dtype)
+    ```
+2. `Layer.call(Ops tf, Operand<T>... inputs)`
+
+    The `call` method transforms input operands, and returns a new tensorflow operand, using
+    tensorflow operations. For example, in a Dense layer, this method runs the computation
+    *f*(*w_k* x + *w_b*).
+
+    As for `build`, the `call` method overridden within a Layer subclass is wrapped by an `apply`
+    method meant for use external to the layer class:
+    ```java
+    // Override this method in a layer subclass
+    protected abstract Operand<T> call(Ops tf, Operand<T>... inputs)
+
+    // Call this method when using a created Layer object externally.
+    // Performs additional checks on the input length / sizes.
+    public final Operand<T> apply(Ops tf, Operand<T>... inputs)
+    ```
+3. `Layer.computeOutputShape(Shape inputShape)`
+
+    Ensuring tensor dimensions are correct through a sequence of transformations
+    is difficult. The `computeOutputShape` method is used to verify that the transformed
+    output of `apply` has the correct shape.
+
+#### Creating Layers
+
+Standard and user-defined layers might require arguments that are optional, and have defaults.
+To allow for this, the standard layer types each define `Options` and `Builder` classes which allow
+the specification of default and optional arguments. Since using an `Options` class for
+every layer can be cumbersome, static helper functions for creating standard layers
+are defined in the `Layers` class.
+
+Creating a `Dense` layer:
+```java
+// Using Layer Options Builder
+Layer<Float> dense = new Dense<>(128, Dense.Options.builder()
```

Comment:
How is the `Float` inferred here? and is `128` a shape? If so, I guess it should be `Shape.make(128)` unless you want to override the `Shape` class in TF Tools. And where is passed the `Ops` instance here, as we saw previously?